### PR TITLE
fix(compiler): preserve this.$event and this.$any accesses in expressions

### LINKED ISSUE #30278: bug(compiler): this.$event and this.$any() should not be allowed in template
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/compiler

### Is this a regression?

 No

### Description

 

Currently `$event` and `$any` are treated as property read of `ImplicitReceiver`(`this`), and expanded to `this.$event`/`this.$any` by parser. Which means that user could manually write `this.$event` to access to `$event` object.

So the the following code are totally valid:

```html
<button (click)="onClick(this.$any(this.$event))">Click Me!</button>
```

It's a really terrible pattern in template usage, gives the impression that these magical properties being assigned to component instance.

Instead, the parser should treat them as lexical identifier in the first place and forbid `this.` usage. It also applies to template reference.

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-mgjryy?file=src%2Fapp%2Fapp.component.html

## 🔥 Exception or Error
<pre><code>N/A</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>7.2.5 now but doesn't matter</code></pre>

**Anything else relevant?**

This should be easy to implement once https://github.com/angular/angular/pull/30277 landed.

### PR DESCRIPTION
**Pull request by crisbeto**
Currently expressions like `$event.foo()` and `this.$event.foo()`, as well as `$any(foo)` and `this.$any(foo)`, are treated as the same expression by the compiler, because `this` is considered the same implicit receiver as when the receiver is omitted. This introduces the following issues:

1. Any time something called `$any` is used, it'll be stripped away, leaving only the first parameter.
2. If something called `$event` is used anywhere in a template, it'll be preserved as `$event`, rather than being rewritten to `ctx.$event`, causing the value to undefined at runtime. This applies to listener, property and text bindings.

These changes resolve the first issue and part of the second one by preserving anything that is accessed through `this`, even if it's one of the "special" ones like `$any` or `$event`. Furthermore, these changes only expose the `$event` global variable inside event listeners, whereas previously it was available everywhere.

Fixes #30278.

### CODE REVIEW

File: packages/compiler/src/expression_parser/ast.ts [hunk 1]
```diff
@@ -85,6 +85,12 @@ export class ImplicitReceiver extends AST {
   }
 }
 
+export class ThisImplicitReceiver extends ImplicitReceiver {
```
Comment (alxhub) [inline packages/compiler/src/expression_parser/ast.ts:88]: This is the only part of this design that I'm concerned about. I can see it roughly mirrors how it was implemented before (where `this.X` was parsed the same as just `X` - producing an AST like `PropertyRead(ImplicitReceiver, 'X')`).

However, I think that really `this` is not an "implicit receiver" - in fact it's very explicitly specified. I'd rather name it `ThisReceiver` and _not_ extend from `ImplicitReceiver`. However, that would require explicitly handling the two the same in downstream consumers (in most cases), so I'm willing to accept the inheritance for now.

Please add documentation to this class explaining how it arises, and maybe a `TODO` to not extend from `ImplicitReceiver` in the future.

File: packages/compiler/src/render3/view/template.ts [hunk 2]
```diff
@@ -1624,10 +1628,16 @@ export class BindingScope implements LocalResolver {
...
-    return new BindingScope().set(0, '$event', o.variable('$event'));
+    return new BindingScope();
   }
 
-  private constructor(public bindingLevel: number = 0, private parent: BindingScope|null = null) {}
+  private constructor(
+      public bindingLevel: number = 0, private parent: BindingScope|null = null,
+      public globals?: Set<string>) {
+    if (globals !== undefined) {
+      globals.forEach(name => this.set(0, name, o.variable(name)));
```
Comment (alxhub) [inline packages/compiler/src/render3/view/template.ts:1638]: Prefer `for..of` loops to `.forEach`.

File: packages/core/test/acceptance/integration_spec.ts [hunk 3]
```diff
@@ -384,6 +384,20 @@ describe('acceptance integration tests', () => {
 
       expect(fixture.nativeElement.innerHTML).toEqual('');
     });
+
+    it('should be able to render the result of a function called $any by using this', () => {
+      @Component({template: '{{this.$any(1)}}'})
```
Comment (alxhub) [inline packages/core/test/acceptance/integration_spec.ts:389]: Maybe also verify in this test that multiple arguments are passed correctly.

**Review by alxhub — APPROVED** [review]

File: packages/compiler/src/expression_parser/ast.ts (hunk 1 above)
Comment (ayazhafiz) [reply packages/compiler/src/expression_parser/ast.ts]: +1 on this, I would also suggest later down the line maybe renaming `ImplicitReceiver` to something like `ThisReceiver` and having a flag on whether it is implicit or not. But this is a breaking change and not worth it anytime soon imo.

**Review by ayazhafiz — APPROVED** [review]
Nice change, LGTM for language service

**Comment by josephperrott** [comment]
@crisbeto Looks like this breaks codelyzer, so it doesn't land cleanly.

While its marked as a fix here, it does modify the type of `AstVisitor` to include a new method.  So everything extending it fails to build.

**Comment by crisbeto** [comment]
@josephperrott I've made the new method optional. Can we give it another try?

**Pull request closed without merging**